### PR TITLE
vpa-recommender: fix history loading by using .QueryRange and use Prometheus client

### LINKED ISSUE #2508: VPA recommender prometheus history provider - cannot get timeseries for cpu: invalid response status: success or type: vector
Prometheus history provider isn't working on VPA 0.6.0+ because the prometheus client expects a matrix for the cpu data but it's a vector as the cpu query now uses rate() as introduced in #1824.

`E1104 04:23:55.334996       1 cluster_feeder.go:213] Cannot get cluster history: cannot get usage history: cannot get timeseries for cpu: invalid response status: success or type: vector
`

### PR DESCRIPTION
**Pull request by milesbxf**
To populate historical data, we'd like to retrieve a set of data points over time from Prometheus. This updates the Prometheus history provider to use the official Prometheus Go client (github.com/prometheus/client_golang). This simplifies the code and means we avoid having to unmarshal into our own custom structs.

This also fixes #2508 by explicitly requesting a range of values over time from Prometheus using the QueryRange API rather than passing in a range vector to Query, which doesn't work for the new rate(cpu[8d]) metric introduced in #1824.

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go [hunk 1]
```diff
@@ -51,54 +51,106 @@ func getDefaultPrometheusHistoryProviderConfigForTest() PrometheusHistoryProvide
 	}
 }
 
-func (m *mockPrometheusClient) GetTimeseries(query string) ([]Timeseries, error) {
-	args := m.Called(query)
-	var returnArg []Timeseries
+type mockPrometheusAPI struct {
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go]: Out of curiosity, is there no mock/fake API provided by the prometheus libs?

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go [hunk 2]
```diff
@@ -51,54 +51,106 @@ func getDefaultPrometheusHistoryProviderConfigForTest() PrometheusHistoryProvide
...
-				"name":      "container"},
-			Samples: []Sample{{
-				Value: 5.5, Timestamp: time.Unix(1, 0)}}}}, nil)
-	mockClient.On("GetTimeseries", memoryQuery).Return([]Timeseries{}, nil)
-	mockClient.On("GetTimeseries", labelsQuery).Return([]Timeseries{}, nil)
+	mockClient.On("QueryRange", mock.Anything, cpuQuery, mock.AnythingOfType("v1.Range")).Return().Return(
+		prommodel.Matrix{
+			{
+				Metric: map[prommodel.LabelName]prommodel.LabelValue{
+
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go]: nit: remove blank line

File: vertical-pod-autoscaler/pkg/recommender/main.go [hunk 3]
```diff
@@ -46,6 +46,7 @@ var (
 	storage = flag.String("storage", "", `Specifies storage mode. Supported values: prometheus, checkpoint (default)`)
 	// prometheus history provider configs
 	historyLength       = flag.String("history-length", "8d", `How much time back prometheus have to be queried to get historical metrics`)
+	historyResolution   = flag.String("history-resolution", "1h", `The interval at which to query historical data`)
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/recommender/main.go]: \`Resolution at which prometheus should be queried for historical metrics\`

File: vertical-pod-autoscaler/pkg/recommender/main.go [hunk 4]
```diff
@@ -93,11 +96,32 @@ func main() {
 }
 
 func createKubeConfig(kubeApiQps float32, kubeApiBurst int) *rest.Config {
```
[the remark this answers is not in the document]
Comment (milesbxf) [reply vertical-pod-autoscaler/pkg/recommender/main.go]: Trying in-cluster then falling back to kubeconfig is a common pattern I see in many tools that interact with Kubernetes - the idea being that a process will only ever be running in a Kubernetes pod or on someone's laptop requiring kubeconfig - they're mutually exclusive.

How would `-kubeconfig` interact with the `KUBECONFIG` environment variable? I feel it's quite important to support that. Then again, this is only really relevant for people hacking on the VPA, so it's probably not that important 😀

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 5]
```diff
@@ -90,7 +102,8 @@ func (p *prometheusHistoryProvider) getContainerIDFromLabels(labels map[string]s
 		ContainerName: containerName}, nil
 }
 
-func (p *prometheusHistoryProvider) getPodIDFromLabels(labels map[string]string) (*model.PodID, error) {
+func (p *prometheusHistoryProvider) getPodIDFromLabels(metric prommodel.Metric) (*model.PodID, error) {
+	labels := promMetricToLabelMap(metric)
```
[the remark this answers is not in the document]
Comment (milesbxf) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: We could, but that'd mean `main.go` would need to convert from the string arg to `prommodel.LabelName`, which is confusing in my opinion. I think we'd also need to marshal all the map lookups from prommodel.LabelValue to string, e.g.:

```
	namespaceValue, ok := labels[p.config.CtrNamespaceLabel]
	if !ok {
		return nil, fmt.Errorf("no %s label", p.config.CtrNamespaceLabel)
	}
	namespace := namespaceValue.(string)
```
which is a bit messy. To be honest the whole `LabelName` / `LabelValue` is a bit of a mistake in the Prometheus library

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 6]
```diff
@@ -125,28 +146,57 @@ func resourceAmountFromValue(value float64, resource model.ResourceName) model.R
...
+		return fmt.Errorf("history length %s is not a valid duration: %v", p.config.HistoryLength, err)
+	}
+
+	end := time.Now()
+	start := end.Add(-time.Duration(historyDuration))
+
+	ctx, cancel := context.WithTimeout(context.Background(), p.queryTimeout)
+	defer cancel()
+
+	step, err := prommodel.ParseDuration(p.config.HistoryResolution)
```
[the remark this answers is not in the document]
Comment (milesbxf) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: Yes, it can - I've updated the comment to clarify

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 7]
```diff
@@ -125,28 +146,57 @@ func resourceAmountFromValue(value float64, resource model.ResourceName) model.R
...
+	if err != nil {
+		return fmt.Errorf("history resolution %s is not a valid duration: %v", p.config.HistoryResolution, err)
+	}
+
+	v, err := p.prometheusClient.QueryRange(ctx, query, prometheusv1.Range{
+		Start: start,
+		End:   end,
+		Step:  time.Duration(step),
```
[the remark this answers is not in the document]
Comment (milesbxf) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: Not at all - this is handled on the Prometheus side. Good question though 👍

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go [hunk 8]
```diff
@@ -160,26 +210,34 @@ func (p *prometheusHistoryProvider) readResourceHistory(res map[model.PodID]*Pod
...
 			res[*podID] = podHistory
 		}
-		podLabels := p.getPodLabelsMap(ts.Labels)
-		for _, sample := range ts.Samples {
-			if sample.Timestamp.After(podHistory.LastSeen) {
-				podHistory.LastSeen = sample.Timestamp
-				podHistory.LastLabels = podLabels
-			}
+		podLabels := p.getPodLabelsMap(ts.Metric)
+		lastSample := ts.Values[len(ts.Values)-1]
```
[the remark this answers is not in the document]
Comment (milesbxf) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider.go]: Yep - I can't find the documentation that explicitly says this, but we always get this in practice. I've added a comment here

File: vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go (hunk 1 above)
Comment (milesbxf) [reply vertical-pod-autoscaler/pkg/recommender/input/history/history_provider_test.go]: Not that I'm aware of

File: vertical-pod-autoscaler/pkg/recommender/main.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bskiba) [reply vertical-pod-autoscaler/pkg/recommender/main.go]: For the ease of review, would you mind splitting this into a separate PR?

File: cluster-autoscaler/cloudprovider/digitalocean/godo/databases_test.go [hunk 9]
```diff
@@ -36,7 +36,7 @@ var db = Database{
 		SSL:      true,
 	},
 	Users: []DatabaseUser{
-		DatabaseUser{
```
Comment (bskiba) [inline cluster-autoscaler/cloudprovider/digitalocean/godo/databases_test.go]: Can you either omit the cluster autoscaler changes or put them in a separate PR?

**Comment by bskiba** [comment]
Can you split the PR as suggested in the comments and then resolve merge conflicts? This will allow me to take another look at the change itself.

**Comment by yashbhutwala** [comment]
@bskiba @milesbxf good stuff here!!  can we pick this up soon?  I can make a PR addressing some of these if need be 😃

**Comment by yashbhutwala** [comment]
/close 

We can close this as it's resolved via #3122 🎉
